File: app/src/main/java/com/kondee/wakeat/MainActivity.java
package com.kondee.wakeat;

import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Parcelable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import com.kondee.wakeat.databinding.ActivityMainBinding;
import com.kondee.wakeat.service.ServiceConstant;

import org.parceler.Parcels;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "Kondee";
    ActivityMainBinding binding;
    private onOptionMenuCreated menuCreatedListener;
    private onOptionItemSelected itemSelectedListener;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);

        initInstance();
    }

    private void initInstance() {

        setSupportActionBar(binding.toolBar);

        Parcelable parcelable = getIntent().getParcelableExtra("parcelable");

        getSupportFragmentManager().beginTransaction()
                .replace(binding.contentContainer.getId(), MainFragment.newInstance(parcelable), "MainFragment")
                .commit();

        Intent intent = new Intent(this, ForegroundLocationService.class);
        intent.setAction(ServiceConstant.STOPFOREGROUND_ACTION);
        startService(intent);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == 0x1 && resultCode == RESULT_OK) {
            if (getSupportFragmentManager().findFragmentById(R.id.contentContainer) instanceof MainFragment) {
                ((MainFragment) getSupportFragmentManager().findFragmentById(R.id.contentContainer)).updateCameraPosition();
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.location, menu);

        if (menuCreatedListener != null) {
            menuCreatedListener.onMenuCreated(menu);
        }
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.location) {
            if (itemSelectedListener != null) {
                itemSelectedListener.onMenuSelected();
            }
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
//        Intent intent = new Intent(MainActivity.this, ForegroundLocationService.class);
//        intent.setAction(ServiceConstant.STARTFOREGROUND_ACTION);
//        startService(intent);
    }

    public interface onOptionMenuCreated {
        void onMenuCreated(Menu menu);
    }

    public void setOnOptionMenuCreatedListener(onOptionMenuCreated menuCreatedListener) {
        this.menuCreatedListener = menuCreatedListener;
    }

    public interface onOptionItemSelected {
        void onMenuSelected();
    }

    public void setOnOptionItemSelectedListener(onOptionItemSelected itemSelectedListener) {
        this.itemSelectedListener = itemSelectedListener;
    }

}

File: app/src/main/java/com/kondee/wakeat/MainActivity.kt
package com.kondee.wakeat

import androidx.appcompat.app.AppCompatActivity
import com.kondee.wakeat.MainActivity.onOptionMenuCreated
import com.kondee.wakeat.MainActivity.onOptionItemSelected
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import com.kondee.wakeat.R
import android.os.Parcelable
import com.kondee.wakeat.MainFragment
import android.content.Intent
import com.kondee.wakeat.ForegroundLocationService
import com.kondee.wakeat.service.ServiceConstant
import android.app.Activity
import android.view.Menu
import android.view.MenuItem
import com.kondee.wakeat.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    var binding: ActivityMainBinding? = null
    private var menuCreatedListener: onOptionMenuCreated? = null
    private var itemSelectedListener: onOptionItemSelected? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        initInstance()
    }

    private fun initInstance() {
        setSupportActionBar(binding!!.toolBar)
        val parcelable = intent.getParcelableExtra<Parcelable>("parcelable")
        supportFragmentManager.beginTransaction()
            .replace(binding!!.contentContainer.id, MainFragment.newInstance(parcelable), "MainFragment")
            .commit()
        val intent = Intent(this, ForegroundLocationService::class.java)
        intent.action = ServiceConstant.STOPFOREGROUND_ACTION
        startService(intent)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 0x1 && resultCode == RESULT_OK) {
            if (supportFragmentManager.findFragmentById(R.id.contentContainer) is MainFragment) {
                (supportFragmentManager.findFragmentById(R.id.contentContainer) as MainFragment?)!!.updateCameraPosition()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.location, menu)
        if (menuCreatedListener != null) {
            menuCreatedListener!!.onMenuCreated(menu)
        }
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.location) {
            if (itemSelectedListener != null) {
                itemSelectedListener!!.onMenuSelected()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    interface onOptionMenuCreated {
        fun onMenuCreated(menu: Menu?)
    }

    fun setOnOptionMenuCreatedListener(menuCreatedListener: onOptionMenuCreated?) {
        this.menuCreatedListener = menuCreatedListener
    }

    interface onOptionItemSelected {
        fun onMenuSelected()
    }

    fun setOnOptionItemSelectedListener(itemSelectedListener: onOptionItemSelected?) {
        this.itemSelectedListener = itemSelectedListener
    }
}